include/pybind11/stl.h

```diff
@@ -36,6 +36,12 @@
 #    include <experimental/optional>
 #    define PYBIND11_HAS_EXP_OPTIONAL 1
 #  endif
+// std::filesystem::path
+#  if defined(PYBIND11_CPP17) && __has_include(<filesystem>) && \
+      PY_VERSION_HEX >= 0x03060000
```

Comment:
Nitpicking: Slightly concerned about this, where the other macros just mention the C++ feature (i.e., `PYBIND11_HAS_OPTIONAL` just means "does C++ has `std::optional`?"). For the future, this has a potential for confusion.

Could/should we split this into two parts, where `PYBIND11_HAS_FILESYSTEM` is here, and the Python check is further down at the definition of the caster? Just a thought; I'm not completely convinced. @bstaletic?

Reply:
Ah, also, could you move this block down 5 lines, please (I said I was nitpicking, didn't I?). I'd like to keep the order `HAS_OPTIONAL`, `HAS_VARIANT`, `HAS_FILESYSTEM` consistent with the next branch of the main `#if`.

Reply:
Done for the second point (and also moved the implementation down to match the order).  I don't really agree with the first point; I think `PYBIND11_HAS_FOO` means "pybind11 supports foo" rather than "C++ has foo" (well, before this PR both matched :)).  In particular changing this would mean that test_stl.cpp would have  to duplicate the version gating logic (which may change if someone decides at some point to also support older pythons).  But I can change that if you really want...

Reply:
> But I can change that if you really want...

I'm not sure if I really want that, either.

I've just been looking at this set of macros like that before (and wonder if in the future, if there's going to be something else depending on `std::filesystem` that's not per se the caster, the same macro might be reused). In principle, we're not advertising that macro, so users shouldn't depend on what exactly it means; in practice, well, ...

Your interpretation does indeed also match the current state, yes, so I don't know. That's why I'm asking other reviewers. :-)

>  In particular changing this would mean that test_stl.cpp would have to duplicate the version gating logic (which may change if someone decides at some point to also support older pythons).

Yes, or another macro that more specifically mentions the caster?

It would even be nicer if there would be an easy way to support 2.7 and 3.5, ofc. But out of scope for this finished PR :-)

Reply:
I'll just leave things as they are for now and see what other reviewers propose.

Reply:
I'd stick with the current definition of the `PYBIND11_HAS_FILESYSTEM` for now. Once pybind11 gets `$NEW_THING` that depends on `<filesystem>`, but not on `os.Path`, the macro can be revised.

Reply:
OK. I was just expecting you to pop up and shout "Hyrum's law", but let's leave it, then ;-)

Reply:
At present, looks good to me! Will mark this convo as resolved.